# Projet TAL - RI
_Charlotte Kruzic & Zoé Marquis_
### Étape 1

In [1]:
!pip install pandas

Defaulting to user installation because normal site-packages is not writeable


In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /home/zoe/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/zoe/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

### Importer les données

In [3]:
df_test = pd.read_csv("test.tsv", sep='\t')
df_train = pd.read_csv("train.tsv", sep='\t')

In [4]:
df_test.head()

category                                           headline  \
0  business  Malick Diouf: incubateurs, startups et entrepr...   
1  business  Pourquoi de nombreuses entreprises américaines...   
2  business  Bronzes du Bénin : 'mon arrière-grand-père a s...   
3  business  Archives géologiques : un milliard d'années on...   
4  business  Des survivants décrivent l'embuscade mortelle ...   

                                                text  \
0  Fondateur de L'africamobile à Dakar, Malick Di...   
1  Un McDonald's à Tampa, en Floride, offre 50 do...   
2  Dans les rues animées de Benin City, au Nigéri...   
3  Par Zaria Gorvett * "FEARFUL DISASTER", titrai...   
4  Les survivants d'une attaque meurtrière contre...   

                                           url  
0  https://www.bbc.com/afrique/region-53099443  
1   https://www.bbc.com/afrique/monde-57127699  
2  https://www.bbc.com/afrique/region-59077317  
3   https://www.bbc.com/afrique/monde-58448245  
4  https://www.bbc.com/afrique/region-50359708

In [5]:
df_train.head()

category                                           headline  \
0    sports         L'Ouganda à l'assaut des "fimbu" de la RDC   
1  business  Stopper la détérioration de l’environnement po...   
2    sports  Coupe d'Afrique des nations 2022 : le sélectio...   
3  business  Tissus africains : pourquoi les teintureries h...   
4  business  Les revenus pendant la pandémie des dix hommes...   

                                                text  \
0  L'Ouganda, placé 79e au classement FIFA le 4 a...   
1  La responsable de la biodiversité des Nations ...   
2  Le sélectionneur de la Sierra Leone, John Keis...   
3  Depuis plus de six siècles, une vaste zone sit...   
4  Pendant la pandémie de coronavirus, la richess...   

                                                 url  
0        https://www.bbc.com/afrique/sports-48730747  
1  https://www.bbc.com/afrique/articles/cevdz4xrp42o  
2        https://www.bbc.com/afrique/region-59881445  
3        https://www.bbc.com/afrique/region-61202726  
4         https://www.bbc.com/afrique/monde-55812018

In [6]:
print(df_train.columns)

Index(['category', 'headline', 'text', 'url'], dtype='object')


## 🧹 Nettoyer les données 

In [7]:
# identifier les valeurs manquantes
valeurs_manquantes = df_train.isnull().sum()
print(valeurs_manquantes) 
# pas de valeur manquante

category    0
headline    0
text        0
url         0
dtype: int64


In [8]:
# identifier les lignes dupliquées
dupliquees = df_train.duplicated()
print(dupliquees.sum())
# 1 ligne dupliquée

1


In [9]:
df_train = df_train.drop_duplicates()
dupliquees = df_train.duplicated()
print(dupliquees.sum())

0


In [10]:
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
Index: 1475 entries, 0 to 1475
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   category  1475 non-null   object
 1   headline  1475 non-null   object
 2   text      1475 non-null   object
 3   url       1475 non-null   object
dtypes: object(4)
memory usage: 57.6+ KB
None


1475 lignes  
toutes les colonnes contiennent du texte  
on s'intéresse seulement aux colonnes `headline` et `text` pour le moment

## Apprentissage de la catégorie
à partir du titre et du texte

In [11]:
X = df_train[['headline', 'text']]
y = df_train.category

In [12]:
X.head()

headline  \
0         L'Ouganda à l'assaut des "fimbu" de la RDC   
1  Stopper la détérioration de l’environnement po...   
2  Coupe d'Afrique des nations 2022 : le sélectio...   
3  Tissus africains : pourquoi les teintureries h...   
4  Les revenus pendant la pandémie des dix hommes...   

                                                text  
0  L'Ouganda, placé 79e au classement FIFA le 4 a...  
1  La responsable de la biodiversité des Nations ...  
2  Le sélectionneur de la Sierra Leone, John Keis...  
3  Depuis plus de six siècles, une vaste zone sit...  
4  Pendant la pandémie de coronavirus, la richess...

In [13]:
y.unique()

array(['sports', 'business', 'politics', 'health', 'technology'],
      dtype=object)

### Validation croisée
découper les données en jeu d'apprentissage et de validation  
20% pour la validation 
avant découpage, on mélange les données
on fixe la graine : random_state=3 pour des fins de reproductibilité

In [14]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=3, shuffle=True)

Estimateurs

Transformateurs

Prédicteurs

Pipeline

`text`

découper le texte en mots
en minuscules
supprimer mots vides

In [22]:
def split_into_tokens_nltk(desc) :
    return word_tokenize(desc)

# Liste des mots vides de NLTK + signes de ponctuation
nltk_stopwords = stopwords.words('french')+list(string.punctuation)

# Objet TfidfVectorizer
desc_vectorizer = TfidfVectorizer(tokenizer=split_into_tokens_nltk,
                                  lowercase=True,
                                  stop_words=nltk_stopwords,
                                  min_df=0.01)

NameError: name 'string' is not defined

## Approche sac de mots

## Approche LSTM bi-directionnel 

## Approche transformers